### ISSUE TITLE
Bitcask fold key issue: binary_to_term(<<>>) [JIRA: RIAK-1751]

### ISSUE THREAD

**Issue by slfritchie** [basho/riak_kv#1031]
(also applies to Riak 2.0?)
Excerpt from http://lists.basho.com/pipermail/riak-users_lists.basho.com/2014-October/016007.html , please refer to it for full context

Hello all,
I am using riak 1.4.3
I have an error during map reduce (not every time) :

{
  "phase": "listkeys",
  "error": "function_clause",
  "input": "xxx",
  "type": "error",
  "stack": "[
    {riak_kv_pipe_listkeys,keysend,
      [ error,
        {worker_crash,{badarg,[{erlang,binary_to_term,[<<>>],[]},
        {riak_kv_bitcask_backend,'-fold_keys_fun/2-fun-1-',4,
                [{file,\"src/riak_kv_bitcask_backend.erl\"},{line,461}]},
         {bitcask_nifs,keydir_fold_cont,4,[{file,\"src/bitcask_nifs.erl\"},{line,419}]},
         {bitcask_nifs,keydir_frozen,4,[{file,\"src/bitcask_nifs.erl\"},{line,265}]},
         {riak_kv_bitcask_backend,'-fold_keys/4-fun-0-',5,
                [{file,\"src/riak_kv_bitcask_backend.erl\"},{line,254}]},
         {riak_kv_worker,handle_work,3,[{file,...},...]},
        ...]},...},...],...},...]"
}

**Comment by slfritchie** [comment]
Reported by @awetzel, I believe.

**Comment by awetzel** [comment]
Yes this issue was reported by me, actually in riak2.0, the key format has changed, but if I look `bk_to_tuple`,
https://github.com/basho/riak_kv/blob/develop/src/riak_kv_bitcask_backend.erl#L110

 I would guess that in the same case in riak2, the error would also appear with a `function_clause bk_to_tuple` since there is no code protecting from a corrupted key.

**Comment by engelsanchez** [comment]
@awetzel Bitcask data records use CRC encoding to verify they are correct and data corrupted on disk should, with all likelihood, be caught before it gets to this point.  But it is possible for a user or script  on the Riak console to add values to the bitcask backend directly with keys that do not conform to what Riak expects. I imagine that is what happened in this case.

We should add protection as a means of extreme paranoia, I don't disagree with that.

**Comment by nickelization** [comment]
This should be easy to fix, and I think I might take a stab at it, but it doesn't seem 100% obvious what the correct behavior should be if we hit on a case like this. Just ignore it? Ignore it and log a warning/error? Should we dump the bad data to a separate location for manual analysis? etc.

If I had to guess, I would probably say ignore the bad data and log a warning, but I'm still pretty new here and am not sure of our policies on logging and such. Figured I'd better check and get some extra opinions.

**Comment by DSomogyi** [comment]
Comment for Jira.

**Comment by engelsanchez** [comment]
Ok. After debugging a related customer issue, I think I finally understand everything that is going on here. I believe this issue points to the root cause of the problem: https://github.com/basho/bitcask/issues/212

On 2.0, reading keys from a corrupted data file when the hintfile is not available or invalid may lead to startup crashes as we try to convert they key. Before 2.0 though, there was no default key transformation in the Riak backend. So the bad empty key would not crash on startup, but ends up loaded into the keydir.  Both the fix to detect corruption properly and wrap calls to convert keys to detect incompatible formats should be added.
